internal/runtime/registry/registry.go

```diff
@@ -0,0 +1,268 @@
+	WarmUp(extensionConfigList *runtimev1.ExtensionConfigList) error
+
+	// IsReady returns true if the RuntimeExtension registry is ready for usage.
+	// This happens after WarmUp is completed.
+	IsReady() bool
+
+	// Add adds all RuntimeExtensions of the given ExtensionConfig.
+	// Please note that if the ExtensionConfig has been added before, the
+	// corresponding registry entries will get updated/replaced with the
+	// one from the newly provided ExtensionConfig.
+	Add(extensionConfig *runtimev1.ExtensionConfig) error
+
+	// Remove removes all RuntimeExtensions corresponding to the provided ExtensionConfig.
+	Remove(extensionConfig *runtimev1.ExtensionConfig) error
+
+	// List all registered RuntimeExtensions for a given catalog.GroupHook.
+	List(gh catalog.GroupHook) ([]*ExtensionRegistration, error)
+
+	// Get the RuntimeExtensions with the given name.
+	Get(name string) (*ExtensionRegistration, error)
+}
+
+// ExtensionRegistration contains information about a registered RuntimeExtension.
+type ExtensionRegistration struct {
+	// Name is the unique name of the RuntimeExtension.
+	Name string
+
+	// ExtensionConfigName is the name of the corresponding ExtensionConfig.
+	ExtensionConfigName string
+
+	// GroupVersionHook is the GroupVersionHook that the RuntimeExtension implements.
+	GroupVersionHook catalog.GroupVersionHook
+
+	// ClientConfig is the ClientConfig to communicate with the RuntimeExtension.
+	ClientConfig runtimev1.ClientConfig
+	// TimeoutSeconds is the timeout duration used for calls to the RuntimeExtension.
+	TimeoutSeconds *int32
+	// FailurePolicy defines how failures in calls to the RuntimeExtension should be handled by a client.
+	FailurePolicy *runtimev1.FailurePolicy
+}
+
+// extensionRegistry is a implementation of ExtensionRegistry.
+type extensionRegistry struct {
+	// ready represents if the registry has been warmed up.
+	ready bool
+	// items contains the registry entries.
+	items map[string]*ExtensionRegistration
+	// lock is used to synchronize access to fields of the extensionRegistry.
+	lock sync.RWMutex
+}
+
+// Extensions returns the ExtensionRegistry singleton.
+func Extensions() ExtensionRegistry {
+	once.Do(func() {
+		instance = extensions()
+	})
+	return instance
+}
+
+func extensions() ExtensionRegistry {
+	return &extensionRegistry{
+		items: map[string]*ExtensionRegistration{},
+	}
+}
+
+// WarmUp can be used to initialize a "cold" RuntimeExtension registry with all
+// known runtimev1.ExtensionConfigs at a given time.
+// After WarmUp completes the RuntimeExtension registry is considered ready.
+func (r *extensionRegistry) WarmUp(extensionConfigList *runtimev1.ExtensionConfigList) error {
+	if extensionConfigList == nil {
+		return errors.New("invalid argument: when calling WarmUp extensionConfigList must not be nil")
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if r.ready {
+		return errors.New("invalid operation: WarmUp cannot be called on a registry which has already been warmed up")
```

Comment:
Q: Will it be a problem to redo warm-up? might be it could be handy in some circumstances...
If case we do it, let's make sure to delete from the registry all the registration for ExtensionConfig that do not exist anymore

Reply:
We can do another warmup but we have to make it very clear that it resets the registry (aka deletes all registrations as you mentioned).

It felt to me like doing the warmup only once is better semantically.